flashlight/fl/tensor/backend/jit/ir/Node.h

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * This source code is licensed under the MIT-style license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#pragma once
+
+#include <list>
+#include <optional>
+#include <vector>
+
+#include "flashlight/fl/tensor/TensorBase.h"
+#include "flashlight/fl/tensor/backend/jit/ir/NodeType.h"
+#include "flashlight/fl/tensor/backend/jit/ir/Use.h"
+
+namespace fl {
+
+using UseList = std::list<Use*>;
+
+/**
+ * A Node represents a step in some Tensor computation.
+ *
+ * Conceptually, nodes form a computation DAG where the result of the
+ * computation is immutable (this must be enforced by all graph rewrites).
+ *
+ * Ownership model:
+ * 1. Node has shared ownership via manual refcount update.
+ * 2. Use is owned by user node.
+ */
+class Node {
+  std::vector<Node*> inputs_;
+  std::vector<UseList::iterator> inputUseIters_;
+  UseList uses_;
+  // Since we enforce dynamic allocation of Nodes via `Subclass::create`, the
+  // refcount always starts from 1 -- representing the `create` function caller.
+  unsigned refCount_{1};
+
+  // present if this node has been evaluated
+  std::optional<Tensor> result_{std::nullopt};
+
+  void nodeImplTypeCheck(NodeType expect, NodeType actual) const;
+
+  // set the input at `inputIdx` -- `resetInput` should be used to clear the old
+  // input at `inputIdx`, if any.
+  void setInputImpl(unsigned inputIdx, Node* input);
+
+  // "unlink" the input at `inputIdx`
+  void resetInput(unsigned inputIdx);
+
+ protected:
+  // A constructor that sets up all the metadata
+  Node(std::vector<Node*>&& inputs);
+
+  // Help enforce internal consistency.
+  Node* getInput(unsigned inputIdx) const;
+
+ public:
+  virtual ~Node();
+
+  // Inputs
+  const std::vector<Node*>& inputs() const;
+  void setInput(unsigned inputIdx, Node* newInput);
+
+  // Uses
+  const UseList& uses() const;
+  void replaceAllUsesWith(Node* newInput);
+
+  // Mainly for debugging/testing
+  unsigned getRefCount() const;
+  // Use carefully -- this manually simulates shared ownership of a node
```

Comment:
why not use `std::shared_ptr`?

Reply:
Yeah I've been struggling with this thing for a long time. Let me document my thought process here:

## Lifetime Management Goals

So my goals for lifetime management here are
1. safe (no memory leak)
2. transparent to the client (as if we have GC)
3. fast (JIT needs to be fast)

## Ownership Requirements
Shared ownership is inevitable in this JIT. There are already these clients that hold nodes for variable amount of time:
1. A `JitTensor` holds a node
2. A `Node` may hold other input nodes, and user nodes.
3. A `GraphVizPrinter` may hold a node for some time, so we can print out not just graphs one by one independently, but also be able to show "shared subgraphs" -- this is particularly helpful for debugging.

## Ownership Implementations
At the very beginning I considered (1). multiple owners (shared_ptr, refcounting, etc.) v.s. (2). a single long-living ownership (like a `Graph` class that holds all nodes).

### Single long-living owner
[PyTorch's JIT](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L92-L95) uses (2), but I kinda don't see any other reason to introduce a `Graph` class, and I _think_ it still requires some ["manual delete"](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/ir/ir.h#L726-L734). So I ditched that idea... Maybe I should look more into that direction?

### Multiple Owners

#### std::shared_ptr
I started with `std::shared_ptr`, ran into issues when I tried to represent `Node-->UserNode` relations, because it creates a cycle with the `Node-->InputNode` relations, and the latter cannot be avoided. However, with the `Use` abstraction now, this actually becomes a nonissue.

So the only concern here is **performance** . But that's just **speculation**, I never measured it.

#### Manual refcount
Basically
1. Restrict ownership sharing to the 3 "big" owners (listed above), i.e., temporary users in places like optimization pass don't share ownership.
2. Manually control shared ownership by using `Node::incRefCount` and `Node::decRefCount`

## Conclusion
1. `std::shared_ptr`
  - __might__ hurt performance
  - slightly more verbose.
2. manual refcount 
  - fast
  - not completely automated

I'm leaning towards (2), and note that for (2), only the shared owners (see above) need to be careful. Ideally future optimization pass writers don't have to worry about this at all -- they just
```c++
oldNode->replaceAllUsesWith(CustomNode::create(...));
```

Reply:
I like this analysis. My thoughts:
1. Agree that an entire `Graph` notion is overkill, and ultimately constrains how different implementers might think about lifetime management
2. `shared_ptr` is annoying for the reasons you mention, and for another: contention with threading for both reads and writes to refcounts, even in places where thread safety might not matter, can be a big problem

The general philosophy with Flashlight and the `fl::Tensor` API thus far has been to automate less, aim for configuration over convention, and prioritize performance and simplicity, so this sounds very in line with that 😄